lib/plug/conn/utils.ex

```diff
@@ -278,24 +281,78 @@ defmodule Plug.Conn.Utils do
+    - `context`: A string providing context about the binary data being checked.
+    - `error_code`: An integer that determines the behavior when invalid UTF-8 is encountered.
+      - `500` will raise the specified `exception` (default).
+      - `404` will return an error tuple.
+      - Codes within `100..999` will log a warning.
+
+  ## Examples
+
+      iex> Plug.Conn.Utils.validate_utf8!(<<255, "invalid">>, RuntimeError, "test context")
+      ** (RuntimeError) invalid UTF-8 on test context, got byte 255 in position 0
+
+      iex> Plug.Conn.Utils.validate_utf8!("valid string", RuntimeError, "test context", 404)
+      :ok
+
+      iex> Plug.Conn.Utils.validate_utf8!(<<255, "invalid">>, RuntimeError, "test context", 404)
+      {:error, "invalid UTF-8 on test context, got byte 255 in position 0"}
+
+
+
+      # Example with logging for invalid UTF-8
+      iex> Plug.Conn.Utils.validate_utf8!(<<255, "invalid">>, RuntimeError, "test context", 200)
+      :ok
+      # Logs "invalid UTF-8 on test context, got byte 255 in position 0"
+
   """
-  @spec validate_utf8!(binary, module, binary) :: :ok | no_return
-  def validate_utf8!(binary, exception, context)
+  @spec validate_utf8!(binary, module, binary, integer()) ::
+          :ok | no_return | {:error, String.t()}
+  def validate_utf8!(binary, exception, context, error_code \\ @utf8_error_code)
+
+  def validate_utf8!(<<binary::binary>>, exception, context, error_code) do
+    do_validate_utf8!(binary, exception, context, error_code, 0)
+  end
 
-  def validate_utf8!(<<binary::binary>>, exception, context) do
-    do_validate_utf8!(binary, exception, context)
+  defp do_validate_utf8!(<<a::56, rest::bits>>, exception, context, error_code, byte_position)
+       when Bitwise.band(a, 0x80808080808080) == 0 do
+    do_validate_utf8!(rest, exception, context, error_code, byte_position + 7)
   end
 
-  defp do_validate_utf8!(<<_::utf8, rest::bits>>, exception, context) do
-    do_validate_utf8!(rest, exception, context)
+  defp do_validate_utf8!(<<_::utf8, rest::bits>>, exception, context, error_code, byte_position) do
+    do_validate_utf8!(rest, exception, context, error_code, byte_position + 1)
   end
 
-  defp do_validate_utf8!(<<byte, _::bits>>, exception, context) do
-    raise exception, "invalid UTF-8 on #{context}, got byte #{byte}"
+  defp do_validate_utf8!(<<byte, _::bits>>, exception, context, error_code, byte_position) do
+    case error_code do
+      500 ->
+        raise exception,
+              "invalid UTF-8 on #{context}, got byte #{byte} in position #{byte_position}"
+
+      404 ->
+        {:error, "invalid UTF-8 on #{context}, got byte #{byte} in position #{byte_position}"}
+
+      error_code when error_code in 100..999 ->
+        :ok =
+          Logger.warn(
+            "Invalid UTF-8 on #{context}, got byte #{byte} in position #{byte_position}",
+            error: @utf8_error_code,
+            context: context,
+            byte: byte
+          )
```

Comment:
It should always raise, and not behave differently based on the format.

Reply:
What information should be in the exception? I guess I'm not exactly clear on what exactly should change with different error codes.

Reply:
If you want a way to not raise, then we should create a `validate_utf8` that returns `{:error, Exception}` without raising. But we should not use the status code to control that. It is unclear what is the ultimate gain here.